discovery-provider/src/tasks/index.py

```diff
@@ -254,13 +256,106 @@ def fetch_tx_receipts(self, block):
     return block_tx_with_receipts
 
 
-def fetch_ipfs_metadata(
+def get_metadata_from_json(default_metadata_fields, resp_json):
+    metadata = {}
+    for parameter, value in default_metadata_fields.items():
+        metadata[parameter] = (
+            resp_json.get(parameter) if resp_json.get(parameter) != None else value
+        )
+    return metadata
+
+
+async def fetch_metadata_from_gateway_endpoints(
+    cid_metadata,
+    cids_txhash_set,
+    cid_to_user_id,
+    user_to_replica_set,
+    cnode_endpoints,
+    cid_type,
+    block_hash,
+    block_number,
+    fetch_from_replica_set=True,
+):
+    async with aiohttp.ClientSession() as async_session:
+        futures = []
+        cid_futures_map = {}
+        cid_txhash_map = {cid: txhash for cid, txhash in cids_txhash_set}
+
+        for cid in cid_txhash_map:
+            if cid in cid_metadata:
+                continue  # already fetched
+            user_id = cid_to_user_id[cid]
+
+            gateway_endpoints = cnode_endpoints
+            if fetch_from_replica_set and user_id and user_id in user_to_replica_set:
+                user_replica_set = user_to_replica_set[
+                    user_id
+                ]  # user or track owner's replica set
+                if not user_replica_set:
+                    continue  # try all gateway endpoints on later attempt
+                gateway_endpoints = user_replica_set.split(",")
+
+            logger.info(f"isaac fetching CID {cid} from {gateway_endpoints}")
+            for gateway_endpoint in gateway_endpoints:
+                future = asyncio.ensure_future(
```

Comment:
we can probably use https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task which is a newer api

Reply:
yeah i think @piazzatron used that. are there any things like this we should add from your changes?

Reply:
i didn't have any changes related to asyncio so whatever you decide. honestly i'm fine with this too. whatever is easiest